src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * Represents a string.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $ignoreCase = false;
+    protected $locale = null;
+
+    abstract public function __construct(string $string = '', string $locale = null);
+
+    public function toGrapheme(): GraphemeString
+    {
+        return new GraphemeString($this->string, $this->locale);
+    }
+
+    public function toUtf8(): Utf8String
+    {
+        return new Utf8String($this->string, $this->locale);
+    }
+
+    public function toBinary(): BinaryString
+    {
+        return new BinaryString($this->string, $this->locale);
+    }
+
+    /**
+     * @return static
+     */
+    public function ignoreCase(): self
+    {
+        $str = clone $this;
+        $str->ignoreCase = true;
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function afterFirst(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function afterLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function append(string $suffix): self;
+
+    /**
+     * @return static
+     */
+    abstract public function beforeFirst(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function beforeLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    abstract public function equalsTo(string $string): bool;
+
+    abstract public function endsWith(string $suffix): bool;
+
+    /**
+     * @return static
+     */
+    public function ensureLeft(string $prefix): self
+    {
+        return $this->startsWith($prefix) ? clone $this : $this->prepend($prefix);
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureRight(string $suffix): self
+    {
+        return $this->endsWith($suffix) ? clone $this : $this->append($suffix);
+    }
+
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    public function isUtf8(): bool
+    {
+        return '' === $this->string || preg_match('//u', $this->string);
+    }
+
+    /**
+     * @return static
+     */
+    public function join(array $strings): self
```

Comment:
>  What about using variadic ? @lyrixx 

not convinced this would be better in practice (and it wouldn't work with associative arrays, making it less useful)

Reply:
Why do you want to pass an associatif array here ? 

More over, variadic will enforce the typehint.

Reply:
I already wrote several pieces of code that implode associative arrays, you can find many in Symfony itself I'm pretty sure. Enforcing the type is not enough a reason here IMHO.

Reply:
I would use variadic as well here. Associative arrays can be transformed upfront if needed.

Reply:
is `$glue->join($parts)` intuitive anyway? :thinking:

Reply:
> is $glue->join($parts) intuitive anyway? thinking

Python has it :)
About variadic, I'm not going to add it. It doesn't make sense to me at all.